packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1918 @@
+   * @returns The internal key of the method.
+   */
+  private getTargetKey(
+    method: ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >['parentCapability'],
+  ): ControllerPermissionSpecification['targetKey'] | undefined {
+    if (hasProperty(this._permissionSpecifications, method)) {
+      return method;
+    }
+
+    const wildCardMethodsWithoutWildCard: Record<string, boolean> = {};
+    for (const targetKey of Object.keys(this._permissionSpecifications)) {
+      const wildCardMatch = targetKey.match(/(.+)\*$/u);
+      if (wildCardMatch) {
+        wildCardMethodsWithoutWildCard[wildCardMatch[1]] = true;
+      }
+    }
+
+    // Check for potentially nested namespaces:
+    // Ex: wildzone_
+    // Ex: eth_plugin_
+    const segments = method.split('_');
+    let targetKey = '';
+
+    while (
+      segments.length > 0 &&
+      !hasProperty(this._permissionSpecifications, targetKey) &&
+      !wildCardMethodsWithoutWildCard[targetKey]
+    ) {
+      targetKey += `${segments.shift()}_`;
+    }
+
+    if (wildCardMethodsWithoutWildCard[targetKey]) {
+      return `${targetKey}*`;
+    }
+
+    return undefined;
+  }
+
+  /**
+   * Grants _approved_ permissions to the specified subject. Every permission and
+   * caveat is stringently validated – including by calling every specification
+   * validator – and an error is thrown if any validation fails.
+   *
+   * **ATTN:** This method does **not** prompt the user for approval.
+   *
+   * @see {@link PermissionController.requestPermissions} For initiating a
+   * permissions request requiring user approval.
+   * @param options - Options bag.
+   * @param options.approvedPermissions - The requested permissions approved by
+   * the user.
+   * @param options.requestData - Permission request data. Passed to permission
+   * factory functions.
+   * @param options.preserveExistingPermissions - Whether to preserve the
+   * subject's existing permissions.
+   * @param options.subject - The subject to grant permissions to.
+   * @returns The granted permissions.
+   */
+  grantPermissions({
+    approvedPermissions,
+    requestData,
+    preserveExistingPermissions = true,
+    subject,
+  }: GrantPermissionsOptions): SubjectPermissions<
+    ExtractPermission<
+      ControllerPermissionSpecification,
+      ControllerCaveatSpecification
+    >
+  > {
+    const { origin } = subject;
+
+    if (!origin || typeof origin !== 'string') {
+      throw new InvalidSubjectIdentifierError(origin);
+    }
+
+    const permissions: SubjectPermissions<
+      ExtractPermission<
+        ControllerPermissionSpecification,
+        ControllerCaveatSpecification
+      >
+    > = ((preserveExistingPermissions && {
+      ...this.getPermissions(origin),
+    }) ||
+      {}) as SubjectPermissions<
+      ExtractPermission<
+        ControllerPermissionSpecification,
+        ControllerCaveatSpecification
+      >
+    >;
```

Comment:
We shouldn't need this cast, nor this explicit type on the declaration, but I can't get this to work without it and I don't understand why.

There's a conflict here between the return type:

```
SubjectPermissions<
  ExtractPermission<
    ControllerPermissionSpecification,
    ControllerCaveatSpecification
  >
>
```

versus the _actual_ type of this value:

```
SubjectPermissions<
  PermissionConstraint & {
    readonly caveats: ExtractCaveats<ControllerCaveatSpecification> extends never
      ? null
      : NonEmptyArray<ExtractCaveats<...>> | null;
    readonly parentCapability: string;
  }
>
```

But... aren't these the same thing? That second type is essentially what `ExtractPermission<...>` is supposed to resolve to.

Reply:
I tried modifying types to get at the core disagreement here, and I think I just found it:

>Type 'SubjectPermissions<ValidPermission<string, ExtractCaveats<ControllerCaveatSpecification>>>' is not assignable to type 'SubjectPermissions<ValidPermission<ControllerPermissionSpecification["targetKey"], ExtractCaveats<ControllerCaveatSpecification>>>'.ts(2322)

Something about how we're using the _state_ type v.s. the return type is coming into conflict. In the state type, the `ControllerPermissionSpecification["targetKey"]` type gets resolved to just `string`. But in the return type, we're expecting not just any string, but a string that happens to be one of the `targetKey` properties of the permission spec.

Reply:
Well, at least we can simplify this a little bit by removing the declared type, as it's redundant if we're casting. Also this seems better expressed with a ternary (which as a bonus, requires one less set of parentheses)

```suggestion
    const permissions = (
      preserveExistingPermissions
        ? {
            ...this.getPermissions(origin),
          }
        : {}
    ) as SubjectPermissions<
      ExtractPermission<
        ControllerPermissionSpecification,
        ControllerCaveatSpecification
      >
    >;
```

Reply:
Hm. Interesting. Smells like index signature shenanigans.

I don't know that it's worth doing anything about at the moment. What do you think?